docking/pose_dock.py

```diff
@@ -119,27 +140,75 @@ def pose_dock(
+                [
+                    np.linspace(max_lambda, 0.0, transition_steps),
+                    np.zeros(n_steps - transition_steps),
+                ]
+            )
+        elif transition_type == "deletion":
+            new_lambda_schedule = np.concatenate(
+                [
+                    np.linspace(0.0, max_lambda, transition_steps),
+                    np.ones(n_steps - transition_steps) * max_lambda,
+                ]
+            )
+
+        calc_work = True
         for step, lamb in enumerate(new_lambda_schedule):
             ctxt.step(lamb)
-            if step % 1000 == 0:
-                print("step", step, "lamb", lamb, "nrg", ctxt.get_u_t())
-                combined_pdb_str = StringIO(Chem.MolToPDBBlock(combined_pdb))
-                pdb_writer = PDBWriter(
-                    combined_pdb_str,
+            if step % 100 == 0 or step == len(new_lambda_schedule) - 1:
+                print(
+                    f"guest_name: {guest_name}\t"
+                    f"step: {str(step).zfill(len(str(n_steps)))}\t"
+                    f"lambda: {lamb:.2f}\t"
+                    f"energy: {ctxt.get_u_t():.2f}"
+                )
+                if ctxt.get_u_t() > 0:
+                    print(f"Error: energy too high for {guest_name}")
+                    calc_work = False
+                    break
+
+                host_coords = ctxt.get_x_t()[: len(host_conf)] * 10
+                host_frame = host_mol.GetConformer()
+                for i in range(host_mol.GetNumAtoms()):
+                    x, y, z = host_coords[i]
+                    host_frame.SetAtomPosition(i, Point3D(x, y, z))
+                conf_id = host_mol.AddConformer(host_frame)
+                writer = PDBWriter(
+                    os.path.join(
+                        outdir,
+                        f"{guest_name}_{str(step).zfill(len(str(n_steps)))}_host.pdb",
+                    )
+                )
+                writer.write(host_mol, conf_id)
+                writer.close()
+
+                guest_coords = ctxt.get_x_t()[len(host_conf) :] * 10
+                guest_frame = guest_mol.GetConformer()
+                for i in range(guest_mol.GetNumAtoms()):
+                    x, y, z = guest_coords[i]
+                    guest_frame.SetAtomPosition(i, Point3D(x, y, z))
+                conf_id = guest_mol.AddConformer(guest_frame)
+                writer = SDWriter(
                     os.path.join(
-                        outdir, f"{guest_name}_{str(step).zfill(len(str(n_steps)))}.pdb"
-                    ),
+                        outdir,
+                        f"{guest_name}_{str(step).zfill(len(str(n_steps)))}_guest.sdf",
+                    )
                 )
-                pdb_writer.write_header()
-                pdb_writer.write(ctxt.get_x_t() * 10)
-                pdb_writer.close()
+                writer.write(guest_mol, conf_id)
+                writer.close()
+
+        if calc_work:
```

Comment:
we should add a sanity check that endpoints in du_dl_obs.full_du_dl() is approximately 0 (eg. absolute value of 1e-3 or less)

Reply:
done